server/src/main/java/org/opensearch/cluster/service/Throttler.java

```diff
@@ -0,0 +1,92 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.cluster.service;
+
+import org.opensearch.common.AdjustableSemaphore;
+
+import java.util.Objects;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+
+/**
+ * Base class for Throttling logic.
+ * It provides throttling functionality over multiple keys.
+ *
+ * @param <T> the type of key on which we want to do throttling.
+ */
+public class Throttler<T> {
+    protected ConcurrentMap<T, AdjustableSemaphore> semaphores = new ConcurrentHashMap<T, AdjustableSemaphore>();
+
+    /**
+     * Method to acquire permits for a key type.
+     * It will return true if permits can be acquired within threshold limits.
+     *
+     * If threshold is not set for key then also it will return True.
+     *
+     * @param type Key for which we want to acquire permits.
+     * @param permits Number of permits to acquire.
+     * @return boolean representing was it able to acquire the permits or not.
+     */
+    public boolean acquire(final T type, final int permits) {
+        assert permits > 0;
+        AdjustableSemaphore semaphore = semaphores.get(type);
+        if (Objects.nonNull(semaphore)) {
```

Comment:
Just curious, why to call a function vs simple check like `semaphore != null` ?

Reply:
It is static function in Objects class, which eventually performs the same check. Added just for readability purpose. 

By taking quick look at the other classes, looks like we check with the operator and not the function for the null check. (`== null`). To keep it consistent, I will change it to `semaphore != null`.